pkg/relayer/miner/miner.go

```diff
@@ -0,0 +1,395 @@
+	eitherClaimedSessions := channel.Map(
+		ctx, sessionsWithOpenClaimWindow,
+		mnr.newMapClaimSessionFn(failedCreateClaimSessionsPublishCh),
+	)
+
+	// TODO_TECHDEBT: pass failed create claim sessions to some retry mechanism.
+	_ = failedCreateClaimSessions
+	logging.LogErrors(ctx, filter.EitherError(ctx, eitherClaimedSessions))
+
+	// Map eitherClaimedSessions to a new observable of relayer.SessionTree which
+	// is notified when the corresponding claim creation succeeded.
+	return filter.EitherSuccess(ctx, eitherClaimedSessions)
+}
+
+// submitProofs maps over the given claimedSessions observable. For each session,
+// it calculates and waits for the earliest block height at which it is safe to
+// submit a proof and does so. It then maps any errors to a new observable which
+// are subsequently logged. It does not block as map operations run in their own
+// goroutines.
+func (mnr *miner) submitProofs(
+	ctx context.Context,
+	claimedSessions observable.Observable[relayer.SessionTree],
+) {
+	// Map claimedSessions to a new observable of the same type which is notified
+	// when the session is eligible to be proven.
+	sessionsWithOpenProofWindow := channel.Map(
+		ctx, claimedSessions,
+		mnr.mapWaitForOpenProofWindow,
+	)
+
+	failedSubmitProofSessions, failedSubmitProveSessionsPublishCh :=
+		channel.NewObservable[relayer.SessionTree]()
+
+	// Map sessionsWithOpenProofWindow to a new observable of an either type,
+	// populated with the session or an error, which is notified after the session
+	// proof has been submitted or an error has been encountered, respectively.
+	eitherProvenSessions := channel.Map(
+		ctx, sessionsWithOpenProofWindow,
+		mnr.newMapProveSessionFn(failedSubmitProveSessionsPublishCh),
+	)
+
+	// TODO_TECHDEBT: pass failed submit proof sessions to some retry mechanism.
+	_ = failedSubmitProofSessions
+	logging.LogErrors(ctx, filter.EitherError(ctx, eitherProvenSessions))
+}
+
+// validateConfigAndSetDefaults ensures that the miner has been configured with
+// a hasher and uses the default hasher if not.
+func (mnr *miner) validateConfigAndSetDefaults() error {
+	if mnr.hasher == nil {
+		mnr.hasher = defaultHasher
+	}
+	return nil
+}
+
+// mapMineRelay is intended to be used as a MapFn. It hashes the relay and compares
+// its difficulty to the minimum threshold. If the relay difficulty is sifficient,
+// it returns an either populated with the minedRelay value. Otherwise, it skips
+// the relay. If it encounters an error, it returns an either populated with the
+// error.
+func (mnr *miner) mapMineRelay(
+	_ context.Context,
+	relay *servicetypes.Relay,
+) (_ either.Either[*minedRelay], skip bool) {
+	relayBz, err := relay.Marshal()
+	if err != nil {
+		return either.Error[*minedRelay](err), true
+	}
+
+	// Is it correct that we need to hash the key while smst.Update() could do it
+	// since smst has a reference to the hasher
+	mnr.hasher.Write(relayBz)
```

Comment:
Nice catch! :raised_hands: 

Instead of synchronizing, I replaced the `hasher.Hasher` with a `func() hasher.Hasher` so that we can construct a new one for each relay and not worry about its state.